src/dumbo/src/tcp/custom_headers.rs

```diff
@@ -0,0 +1,158 @@
+// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use micro_http::{HttpHeaderError, RequestError};
+use std::collections::HashMap;
+use std::result::Result;
+
+/// Wrapper over supported HTTP Custom Header types.
+#[derive(Debug, Eq, Hash, PartialEq)]
+pub enum CustomHeader {
+    /// Header `X-aws-ec2-metadata-token`
+    XAwsMetadataToken,
+    /// Header `X-aws-ec2-metadata-token-ttl-seconds`
+    XAwsMetadataTokenTtlSeconds,
+}
+
+impl CustomHeader {
+    /// Parses a byte slice into a `CustomHeader` structure.
+    fn try_from(mut string: String) -> Result<Self, RequestError> {
+        string.make_ascii_lowercase();
+        match string.trim() {
+            "x-aws-ec2-metadata-token" => Ok(Self::XAwsMetadataToken),
+            "x-aws-ec2-metadata-token-ttl-seconds" => Ok(Self::XAwsMetadataTokenTtlSeconds),
+            invalid_key => Err(RequestError::HeaderError(HttpHeaderError::UnsupportedName(
+                invalid_key.to_string(),
+            ))),
+        }
+    }
+}
+
+/// Wrapper over the list of custom headers associated with a Request.
+#[derive(Debug, PartialEq)]
+pub struct CustomHeaders {
```

Comment:
I thought about restructuring the `tcp` module, too, as it was a little confusing to me when I first dove into it. However, I didn't want to burden this PR with a reorganization, as it already is pretty dense and on top of these changes I will also add some integration tests. I think it would be very hard to review. 

How about I treat this as a technical debt and post a PR with the restructured `dumbo/src/tcp` module once we merge this one? WDYT?

Reply:
That's fine, but it might not be necessary depending on what we decide regarding this: https://github.com/firecracker-microvm/firecracker/pull/2643#discussion_r665337894

My point being that `CustomHeaders` is mmds-logic-specific and are only used in mmds crate and would be better suited in `mmds` crate rather than `dumbo` crate.

Reply:
See latest https://github.com/firecracker-microvm/firecracker/pull/2643#discussion_r666019702